numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -3692,9 +3691,12 @@ static void
         }
     }
     else {
-        for (i = 0; i < ni; i++) {
+        for (i = 0, j = 0; i < ni; i++, j++) {
+            if (j >= nv) {
+                j = 0;
+            }
```

Comment:
The extra performance was of course coming from leaving the mask check out... Putting it back in the following takes the same time (1.55 us) as the much more readable single if in the current PR:

```
for (i = 0; i < ni;) {
    for (j = 0; j < nv && i < ni; ++j, ++i) {
        if (mask[i]) {
            in[i] = vals[j];
        }
    }
}
```